File: apps/api/app/api/workflows.py
from datetime import UTC, datetime

from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.agents.orchestrator import AgentOrchestrator
from app.core.exceptions import NotFoundError
from app.core.security import get_current_user
from app.core.workflow_engine import execute_workflow
from app.database import get_db
from app.models import User
from app.models.workflow import RunStatus, Workflow, WorkflowRun
from app.schemas.workflow import (
    WorkflowCreate,
    WorkflowExecuteRequest,
    WorkflowResponse,
    WorkflowRunResponse,
    WorkflowUpdate,
)

router = APIRouter(prefix="/workflows", tags=["workflows"])

_orchestrator: AgentOrchestrator | None = None


def _get_orch() -> AgentOrchestrator:
    global _orchestrator
    if _orchestrator is None:
        _orchestrator = AgentOrchestrator()
    return _orchestrator


# ── Workflows CRUD ──────────────────────────────────────────────────

@router.get("", response_model=list[WorkflowResponse])
async def list_workflows(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.execute(
        select(Workflow)
        .where(Workflow.owner_id == current_user.id)
        .order_by(Workflow.updated_at.desc())
    )
    return [WorkflowResponse.model_validate(w) for w in result.scalars().all()]


@router.post("", response_model=WorkflowResponse, status_code=201)
async def create_workflow(
    data: WorkflowCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    wf = Workflow(
        name=data.name,
        description=data.description,
        nodes=data.nodes,
        edges=data.edges,
        owner_id=current_user.id,
    )
    db.add(wf)
    await db.commit()
    await db.refresh(wf)
    return WorkflowResponse.model_validate(wf)


@router.get("/{workflow_id}", response_model=WorkflowResponse)
async def get_workflow(
    workflow_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.execute(
        select(Workflow).where(Workflow.id == workflow_id, Workflow.owner_id == current_user.id)
    )
    wf = result.scalar_one_or_none()
    if not wf:
        raise NotFoundError("Workflow not found")
    return WorkflowResponse.model_validate(wf)


@router.put("/{workflow_id}", response_model=WorkflowResponse)
async def update_workflow(
    workflow_id: int,
    data: WorkflowUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.execute(
        select(Workflow).where(Workflow.id == workflow_id, Workflow.owner_id == current_user.id)
    )
    wf = result.scalar_one_or_none()
    if not wf:
        raise NotFoundError("Workflow not found")

    update_data = data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(wf, field, value)
    wf.updated_at = datetime.now(UTC)
    await db.commit()
    await db.refresh(wf)
    return WorkflowResponse.model_validate(wf)


@router.delete("/{workflow_id}", status_code=204)
async def delete_workflow(
    workflow_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.execute(
        select(Workflow).where(Workflow.id == workflow_id, Workflow.owner_id == current_user.id)
    )
    wf = result.scalar_one_or_none()
    if not wf:
        raise NotFoundError("Workflow not found")
    await db.delete(wf)
    await db.commit()


# ── Execution ───────────────────────────────────────────────────────

@router.post("/{workflow_id}/execute", response_model=WorkflowRunResponse)
async def execute_workflow_endpoint(
    workflow_id: int,
    body: WorkflowExecuteRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.execute(
        select(Workflow).where(Workflow.id == workflow_id, Workflow.owner_id == current_user.id)
    )
    wf = result.scalar_one_or_none()
    if not wf:
        raise NotFoundError("Workflow not found")

    run = WorkflowRun(
        workflow_id=wf.id,
        trigger=body.trigger,
        input_data=body.input_data,
    )
    db.add(run)
    await db.commit()
    await db.refresh(run)

    try:
        await execute_workflow(wf, run, orchestrator=_get_orch())
    except Exception as e:
        run.status = RunStatus.FAILED
        run.error = str(e)
        run.completed_at = datetime.now(UTC)

    await db.commit()
    await db.refresh(run)
    return WorkflowRunResponse.model_validate(run)


# ── Runs ────────────────────────────────────────────────────────────

@router.get("/{workflow_id}/runs", response_model=list[WorkflowRunResponse])
async def list_runs(
    workflow_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.execute(
        select(WorkflowRun)
        .where(WorkflowRun.workflow_id == workflow_id)
        .order_by(WorkflowRun.created_at.desc())
        .limit(50)
    )
    return [WorkflowRunResponse.model_validate(r) for r in result.scalars().all()]


@router.get("/runs/{run_id}", response_model=WorkflowRunResponse)
async def get_run(
    run_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.execute(select(WorkflowRun).where(WorkflowRun.id == run_id))
    run = result.scalar_one_or_none()
    if not run:
        raise NotFoundError("Run not found")
    return WorkflowRunResponse.model_validate(run)


@router.get("/runs", response_model=list[WorkflowRunResponse])
async def list_all_runs(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.execute(
        select(WorkflowRun)
        .join(Workflow)
        .where(Workflow.owner_id == current_user.id)
        .order_by(WorkflowRun.created_at.desc())
        .limit(100)
    )
    return [WorkflowRunResponse.model_validate(r) for r in result.scalars().all()]


@router.post("/runs/{run_id}/cancel", response_model=WorkflowRunResponse)
async def cancel_run(
    run_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    result = await db.execute(select(WorkflowRun).where(WorkflowRun.id == run_id))
    run = result.scalar_one_or_none()
    if not run:
        raise NotFoundError("Run not found")
    run.status = RunStatus.CANCELLED
    run.completed_at = datetime.now(UTC)
    await db.commit()
    await db.refresh(run)
    return WorkflowRunResponse.model_validate(run)

File: apps/web/src/stores/workflows.ts
import { create } from "zustand"
import { api } from "@/lib/api"

export interface WorkflowNode {
  id: string
  type: string
  position: { x: number; y: number }
  data: Record<string, unknown>
}

export interface WorkflowEdge {
  id: string
  source: string
  target: string
  sourceHandle?: string
  targetHandle?: string
  label?: string
}

export interface Workflow {
  id: number
  name: string
  description: string | null
  status: string
  nodes: WorkflowNode[]
  edges: WorkflowEdge[]
  owner_id: number
  created_at: string
  updated_at: string
}

export interface WorkflowRun {
  id: number
  workflow_id: number
  status: string
  trigger: string
  input_data: Record<string, unknown>
  output_data: Record<string, unknown> | null
  logs: WorkflowLog[]
  error: string | null
  started_at: string | null
  completed_at: string | null
  created_at: string
}

export interface WorkflowLog {
  node_id: string
  node_type: string
  status: string
  timestamp: string
  result?: unknown
  error?: string
}

interface WorkflowState {
  workflows: Workflow[]
  runs: WorkflowRun[]
  current: Workflow | null
  loading: boolean
  error: string | null
  fetch: () => Promise<void>
  get: (id: number) => Promise<Workflow>
  create: (data: { name: string; description?: string }) => Promise<Workflow>
  save: (id: number, data: Partial<Workflow>) => Promise<Workflow>
  remove: (id: number) => Promise<void>
  execute: (id: number, input?: Record<string, unknown>) => Promise<WorkflowRun>
  fetchRuns: (workflowId?: number) => Promise<void>
  getRun: (runId: number) => Promise<WorkflowRun>
  cancelRun: (runId: number) => Promise<WorkflowRun>
  setCurrent: (wf: Workflow | null) => void
}

export const useWorkflowStore = create<WorkflowState>((set, get) => ({
  workflows: [],
  runs: [],
  current: null,
  loading: false,
  error: null,

  fetch: async () => {
    set({ loading: true, error: null })
    try {
      const workflows = await api.get("/workflows")
      set({ workflows, loading: false })
    } catch {
      set({ loading: false, error: "Failed to load workflows" })
    }
  },

  get: async (id) => {
    const wf = await api.get(`/workflows/${id}`)
    set({ current: wf })
    return wf
  },

  create: async (data) => {
    const wf = await api.post("/workflows", data)
    set({ workflows: [wf, ...get().workflows] })
    return wf
  },

  save: async (id, data) => {
    const updated = await api.put(`/workflows/${id}`, data)
    set({ current: updated, workflows: get().workflows.map((w) => (w.id === id ? updated : w)) })
    return updated
  },

  remove: async (id) => {
    await api.delete(`/workflows/${id}`)
    set({ workflows: get().workflows.filter((w) => w.id !== id) })
  },

  execute: async (id, input = {}) => {
    return api.post(`/workflows/${id}/execute`, { input_data: input })
  },

  fetchRuns: async (workflowId) => {
    set({ loading: true })
    try {
      const path = workflowId ? `/workflows/${workflowId}/runs` : "/workflows/runs"
      const runs = await api.get(path)
      set({ runs, loading: false })
    } catch {
      set({ loading: false, error: "Failed to load runs" })
    }
  },

  getRun: async (runId) => {
    return api.get(`/workflows/runs/${runId}`)
  },

  cancelRun: async (runId) => {
    return api.post(`/workflows/runs/${runId}/cancel`)
  },

  setCurrent: (wf) => set({ current: wf }),
}))
